services/topology-engine/queue-engine/topologylistener/eventhandler.py

```diff
@@ -13,68 +13,110 @@
 #   limitations under the License.
 #
 
-import kafkareader
 import json
+import logging
 import time
 
 import gevent
 import gevent.pool
 import gevent.queue
+import kafkareader
 
-from messageclasses import MessageItem
-import logging
-import config
-
-logger = logging.getLogger(__name__)
+from topologylistener import config
+from topologylistener import const
+from topologylistener import context as context_module
+from topologylistener import exc
+from topologylistener.messageclasses import MessageItem
 
-known_messages = ['org.openkilda.messaging.info.event.SwitchInfoData',
-                  'org.openkilda.messaging.info.event.IslInfoData',
-                  'org.openkilda.messaging.info.event.PortInfoData',
-                  'org.openkilda.messaging.info.flow.FlowInfoData']
-known_commands = ['org.openkilda.messaging.command.flow.FlowCreateRequest',
-                  'org.openkilda.messaging.command.flow.FlowDeleteRequest',
-                  'org.openkilda.messaging.command.flow.FlowUpdateRequest',
-                  'org.openkilda.messaging.command.flow.FlowPathRequest',
-                  'org.openkilda.messaging.command.flow.FlowGetRequest',
-                  'org.openkilda.messaging.command.flow.FlowsGetRequest',
-                  'org.openkilda.messaging.command.flow.FlowRerouteRequest',
-                  'org.openkilda.messaging.command.discovery.NetworkCommandData']
+raw_logger = logging.getLogger(__name__)
 
 
 def main_loop():
     pool_size = config.getint('gevent', 'worker.pool.size')
     pool = gevent.pool.Pool(pool_size)
-    logger.info('Started gevent pool with size %d', pool_size)
+    raw_logger.info('Started gevent pool with size %d', pool_size)
 
     consumer = kafkareader.create_consumer(config)
 
+    raw_logger.info('Ready to handle requests')
+
     while True:
         try:
             raw_event = kafkareader.read_message(consumer)
-            logger.debug('READ MESSAGE %s', raw_event)
-            event = MessageItem(**json.loads(raw_event))
-
-            if event.get_message_type() in known_messages\
-                    or event.get_command() in known_commands:
-                pool.spawn(topology_event_handler, event)
-            else:
-                logger.debug('Received unknown type or command %s', raw_event)
+            context, request = unpack(raw_event)
+            context.log(raw_logger).debug(
+                    'Raw request payload (see extra field %s)',
+                    const.LOG_ATTR_JSON_PAYLOAD,
+                    extra={const.LOG_ATTR_JSON_PAYLOAD: request})
 
+            handler = gevent.Greenlet(event_handler, context, request)
```

Comment:
This is the same what pool.spawn do.

I need to register(handler.link) a callable that will be called after handler do it's job. It can be done and with spawn `pool.spawn(...).link(...)`, but in this case handler start inside spawn and can end before .link() call... this can lead to "skip" call of callable passed in .link(...)